src/primitives/sinsemilla.rs

```diff
@@ -64,59 +61,124 @@ impl<I: Iterator<Item = bool>> Iterator for Pad<I> {
+/// [`sinsemilla::short_commit`] can be used.
 ///
-/// [concretesinsemillacommit]: https://zips.z.cash/protocol/nu5.pdf#concretesinsemillacommit
+/// [`sinsemilla::commit`]: self::commit
+/// [`sinsemilla::short_commit`]: self::short_commit
 #[allow(non_snake_case)]
-pub(crate) fn commit(
-    domain_prefix: &str,
-    msg: impl Iterator<Item = bool>,
-    r: &pallas::Scalar,
-) -> pallas::Point {
-    let m_prefix = domain_prefix.to_owned() + "-M";
-    let r_prefix = domain_prefix.to_owned() + "-r";
+pub struct CommitDomain {
+    Q: pallas::Point,
+    R: pallas::Point,
+}
 
-    let hasher_r = pallas::Point::hash_to_curve(&r_prefix);
+impl CommitDomain {
+    /// Construct a new CommitDomain with a specific prefix string
+    pub(crate) fn new(domain: &str) -> Self {
+        let m_prefix = domain.to_owned() + "-M";
+        let r_prefix = domain.to_owned() + "-r";
+        CommitDomain {
+            Q: HashDomain::new(&m_prefix).Q,
+            R: pallas::Point::hash_to_curve(&r_prefix.clone())(&[]),
+        }
+    }
 
-    hash_to_point(&m_prefix, msg) + hasher_r(&[]) * r
-}
+    /// `SinsemillaHashToPoint` from [§ 5.4.1.9][concretesinsemillahash].
+    ///
+    /// [concretesinsemillahash]: https://zips.z.cash/protocol/nu5.pdf#concretesinsemillahash
+    #[allow(non_snake_case)]
+    pub(crate) fn hash_to_point(&self, msg: impl Iterator<Item = bool>) -> pallas::Point {
+        let padded: Vec<_> = Pad::new(msg).collect();
 
-/// `SinsemillaShortCommit` from [§ 5.4.7.4][concretesinsemillacommit].
-///
-/// [concretesinsemillacommit]: https://zips.z.cash/protocol/nu5.pdf#concretesinsemillacommit
-pub(crate) fn short_commit(
-    domain_prefix: &str,
-    msg: impl Iterator<Item = bool>,
-    r: &pallas::Scalar,
-) -> pallas::Base {
-    extract_p(&commit(domain_prefix, msg, r))
+        let hasher_S = pallas::Point::hash_to_curve(S_PERSONALIZATION);
+        let S = |chunk: &[bool]| hasher_S(&lebs2ip_k(chunk).to_le_bytes());
+
+        padded
+            .chunks(K)
+            .fold(self.Q, |acc, chunk| acc.double() + S(chunk))
+    }
+
+    /// `SinsemillaCommit` from [§ 5.4.7.4][concretesinsemillacommit].
+    ///
+    /// [concretesinsemillacommit]: https://zips.z.cash/protocol/nu5.pdf#concretesinsemillacommit
+    #[allow(non_snake_case)]
+    pub(crate) fn commit(
+        &self,
+        msg: impl Iterator<Item = bool>,
+        r: &pallas::Scalar,
+    ) -> pallas::Point {
+        self.hash_to_point(msg) + self.R * r
+    }
+
+    /// `SinsemillaShortCommit` from [§ 5.4.7.4][concretesinsemillacommit].
+    ///
+    /// [concretesinsemillacommit]: https://zips.z.cash/protocol/nu5.pdf#concretesinsemillacommit
+    pub(crate) fn short_commit(
+        &self,
+        msg: impl Iterator<Item = bool>,
+        r: &pallas::Scalar,
+    ) -> pallas::Base {
+        extract_p(&self.commit(msg, r))
+    }
+
+    /// Return `Q`
+    #[allow(non_snake_case)]
+    pub(crate) fn Q(&self) -> pallas::Point {
```

Comment:
Where will this be used? (Same question for the other unused accessors).

Reply:
I'm using `R()` in tests for Orchard constants:
- https://github.com/zcash/orchard/pull/43/files#diff-67648b0a12585bf10d8511f896af5c7226e18f770cf74cfb8aace4af187766aaR43
- https://github.com/zcash/orchard/pull/43/files#diff-0236c6fdfb7dc339abd6f277fab892bc8a17a158107dd164a5ba7e330348fdb8R43

Reply:
If these are only used in tests, then place them behind `#[cfg(test)]`.